File: rd-framework/src/main/kotlin/com/jetbrains/rider/framework/impl/RdSignal.kt
package com.jetbrains.rider.framework.impl

import com.jetbrains.rider.framework.ISerializer
import com.jetbrains.rider.framework.Polymorphic
import com.jetbrains.rider.framework.SerializationCtx
import com.jetbrains.rider.framework.AbstractBuffer
import com.jetbrains.rider.framework.base.RdReactiveBase
import com.jetbrains.rider.framework.base.withId
import com.jetbrains.rider.framework.readRdId
import com.jetbrains.rider.util.lifetime.Lifetime
import com.jetbrains.rider.util.reactive.*
import com.jetbrains.rider.util.string.printToString
import com.jetbrains.rider.util.trace

class RdSignal<T>(val valueSerializer: ISerializer<T> = Polymorphic<T>()) : RdReactiveBase(), IAsyncSignal<T> {

    companion object : ISerializer<RdSignal<*>>{
        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdSignal<*> = read(ctx, buffer, Polymorphic<Any?>())
        fun<T> read(ctx: SerializationCtx, buffer: AbstractBuffer, valueSerializer: ISerializer<T>): RdSignal<T> = RdSignal(valueSerializer).withId(buffer.readRdId())
        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdSignal<*>) = value.rdid.write(buffer)
    }

    private val signal = Signal<T>()
    override val changing: Boolean get() = signal.changing

    override lateinit var wireScheduler: IScheduler
    override lateinit var serializationContext : SerializationCtx


    override fun onWireReceived(buffer: AbstractBuffer) {
        val value = valueSerializer.read(serializationContext, buffer)
        logReceived.trace {"signal `$location` ($rdid):: value = ${value.printToString()}"}
        signal.fire(value)
    }

    //protocol init, don't mess with initializer above
    override fun init(lifetime: Lifetime) {
        super.init(lifetime)
        serializationContext = super.serializationContext
        wireScheduler = defaultScheduler
        wire.advise(lifetime, this)

    }

    override fun fire(value: T) {
//        assertBound()
        if (!async) assertThreading()
        //localChange {
        wire.send(rdid) { buffer ->
            logSend.trace {"signal `$location` ($rdid):: value = ${value.printToString()}"}
            valueSerializer.write(serializationContext, buffer, value)
        }
        signal.fire(value)
        //}
    }


    override fun advise(lifetime: Lifetime, handler: (T) -> Unit) {
        if (isBound) assertThreading()
        signal.advise(lifetime, handler)
    }


    override fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (T) -> Unit) {
        if (isBound) assertThreading() //even if listener on pool thread, advise must be on main thread
        this.wireScheduler = scheduler
        signal.advise(lifetime, handler)
    }
}

File: rd-net/RdFramework/Impl/RdSignal.cs
using System;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Serialization;
using JetBrains.Threading;

namespace JetBrains.Rd.Impl
{
  public class RdSignal<T> : RdReactiveBase, ISignal<T>
  {
    #region Serializers

    public CtxReadDelegate<T> ReadValueDelegate { get; }
    public CtxWriteDelegate<T> WriteValueDelegate { get; }

    public static RdSignal<T> Read(SerializationCtx ctx, UnsafeReader reader)
    {
      return Read(ctx, reader, Polymorphic<T>.Read, Polymorphic<T>.Write);
    }

    public static RdSignal<T> Read(SerializationCtx _, UnsafeReader reader, CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
    {
      var id = reader.ReadRdId();
      return new RdSignal<T>(readValue, writeValue).WithId(id);
    }

    public static void Write(SerializationCtx ctx, UnsafeWriter writer, RdSignal<T> value)
    {
      writer.Write(value.RdId);
    }

    #endregion

    private readonly Signal<T> mySignal = new();

    public IScheduler? Scheduler { get; set; }


    public RdSignal(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
    {
      ReadValueDelegate = readValue;
      WriteValueDelegate = writeValue;
    }

    protected override void PreInit(Lifetime lifetime, IProtocol parentProto)
    {
      base.PreInit(lifetime, parentProto);

      parentProto.Wire.Advise(lifetime, this);
    }

    public override RdWireableContinuation OnWireReceived(Lifetime lifetime, IProtocol proto, SerializationCtx ctx, UnsafeReader reader, UnsynchronizedConcurrentAccessDetector? _)
    {
      var value = ReadValueDelegate(ctx, reader);
      ReceiveTrace?.Log($"{this} :: value = {value.PrintToString()}");
      var scheduler = Scheduler ?? proto.Scheduler;
      return new RdWireableContinuation(lifetime, scheduler, null, () =>
      {
        using (UsingDebugInfo())
          mySignal.Fire(value);
      });
    }

    public void Fire(T value)
    {
      AssertThreading();
      
      var proto = TryGetProto();
      if (!Async || proto is { Scheduler.IsActive: true })
        AssertBound();

      AssertNullability(value);

      var wire = proto?.Wire;
      if (wire == null && Async)
        return;
      
      if (!TryGetSerializationContext(out var serializationCtx))
        return;

      //local change
      wire.NotNull(this).Send(RdId, SendContext.Of(serializationCtx, value, this), static (sendContext, stream) =>
      {
        var me = sendContext.This;
        SendTrace?.Log($"{me} :: value = {sendContext.Event.PrintToString()}");

        me.WriteValueDelegate(sendContext.SzrCtx, stream, sendContext.Event);
      });
      
      using (UsingDebugInfo())
        mySignal.Fire(value);
    }


    public void Advise(Lifetime lifetime, Action<T> handler)
    {
      mySignal.Advise(lifetime, handler);
      
      if (IsBound)
        AssertThreading();
    }

    protected override string ShortName => "signal";
  }
}